include/wil/com.h

```diff
@@ -1984,6 +1988,140 @@ namespace wil
     {
         return CoGetClassObjectNoThrow<Interface>(__uuidof(Class), dwClsContext);
     }
+
+#if __has_include(<tuple>) && (__WI_LIBCPP_STD_VER >= 17)
+    namespace details
+    {
+        template <typename error_policy, typename... Results>
+        auto CoCreateInstanceEx(REFCLSID clsid, CLSCTX clsCtx) noexcept
+        {
+            MULTI_QI multiQis[sizeof...(Results)]{};
+            const IID* iids[sizeof...(Results)]{ &__uuidof(Results)... };
+
+            static_assert(sizeof...(Results) > 0);
+
+            for (auto i = 0U; i < sizeof...(Results); ++i)
+            {
+                multiQis[i].pIID = iids[i];
+            }
+
+            const auto hr = CoCreateInstanceEx(clsid, nullptr, clsCtx, nullptr,
+                ARRAYSIZE(multiQis), multiQis);
+
+            std::tuple<wil::com_ptr_t<Results, error_policy>...> resultTuple;
+
+            std::apply([i = 0, &multiQis](auto&... a) mutable
+            {
+                (a.attach(reinterpret_cast<typename std::remove_reference_t<decltype(a)>::pointer>(multiQis[i++].pItf)), ...);
+            }, resultTuple);
+            return std::tuple<HRESULT, decltype(resultTuple)>(hr, std::move(resultTuple));
+        }
+
+        template<typename error_policy, typename... Results>
+        auto com_multi_query(IUnknown* obj)
+        {
+            MULTI_QI multiQis[sizeof...(Results)]{};
+            const IID* iids[sizeof...(Results)]{ &__uuidof(Results)... };
+
+            static_assert(sizeof...(Results) > 0);
+
+            for (auto i = 0U; i < sizeof...(Results); ++i)
+            {
+                multiQis[i].pIID = iids[i];
+            }
+
+            std::tuple<wil::com_ptr_t<Results, error_policy>...> resultTuple{};
+
+            wil::com_ptr_nothrow<IMultiQI> multiQi;
+            auto hr = obj->QueryInterface(IID_PPV_ARGS(&multiQi));
+            if (SUCCEEDED(hr))
+            {
+                hr = multiQi->QueryMultipleInterfaces(ARRAYSIZE(multiQis), multiQis);
+                std::apply([i = 0, &multiQis](auto&... a) mutable
+                {
+                    (a.attach(reinterpret_cast<typename std::remove_reference_t<decltype(a)>::pointer>(multiQis[i++].pItf)), ...);
+                }, resultTuple);
+            }
+            return std::tuple<HRESULT, decltype(resultTuple)>{hr, std::move(resultTuple)};

```

Comment:
Why not use std::tuple_cat?

Reply:
I'll assume that the idea was to return a tuple of size 2. Personally, I'd prefer a flat tuple so that it works better with structured bindings, but what's done is done and I'm not going to change the function signature here.

Reply:
I suppose it also makes it easier to implement the throwing/failfast variants this way

Reply:
I'm also being told that @oldnewthing is making changes to this code, but I'm unsure of what those are right now